torch/_dynamo/backends/distributed.py

```diff
@@ -329,27 +313,47 @@ def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
         debug_str += "\n---------------\n"
         ddp_graph_log.debug(debug_str)
 
-        # 3: compile each of the partitioned submodules using the user-provided compiler
+        # 3: create a wrapper which lazily compiles each of the partitioned submodules
+        # using the user-provided compiler
         class SubmodCompiler(torch.fx.interpreter.Interpreter):
             def __init__(self, module, compiler):
                 super().__init__(module)
                 self.compiler = compiler
 
-            def compile_submod(self, input_mod, args, kwargs):
+            def lazily_compile_submod(self, input_mod):
                 """
                 Compile the submodule,
                 using a wrapper to make sure its output is always a tuple,
                 which is required by AotAutograd based compilers
                 """
-                assert len(kwargs) == 0, "We assume only args for these modules"
 
                 class WrapperModule(torch.nn.Module):
-                    def __init__(self, submod, unwrap_singleton_tuple):
+                    def __init__(self, submod, compiler, unwrap_singleton_tuple):
                         super().__init__()
                         self.submod = submod
+                        self.compiler = compiler
+                        self.compiled = False
                         self.unwrap_singleton_tuple = unwrap_singleton_tuple
 
                     def forward(self, *args):
+                        if not self.compiled:
+                            fake_args = []
+                            for arg in args:
+                                if isinstance(arg, torch.Tensor) and not is_fake(arg):
+                                    fake_args.append(
+                                        torch._dynamo.utils.to_fake_tensor(
+                                            arg, fake_mode
+                                        )
+                                    )
+                                else:
+                                    fake_args.append(arg)
+                            # First trace with fake args
```

Comment:
is it actually necessary to do the first trace with fake args?

IIRC it may have been necessary at one point, but i was trying something similar recently and it seemed like maybe this requirement is relaxed now?

Reply:
If you run with real args, it will run the real computation, which may include mutations.

Reply:
yea, you're right.